concepts/eventing/performance evaluation/nats-jetstream/results_prod_HA.md

```diff
@@ -0,0 +1,332 @@
+
+## NATS Pods Anti-Affinity
+
+The NATS pods have the AntiAffinity ([here](https://github.com/kyma-project/kyma/blob/6eba6d3b62c15c69645f8577d5c8ca46beeb2c93/resources/eventing/charts/nats/values.yaml#L144)) :
+```
+affinity:
+  podAntiAffinity:
+    preferredDuringSchedulingIgnoredDuringExecution:
+    - podAffinityTerm:
+        labelSelector:
+          matchLabels:
+            nats_cluster: eventing-nats
+        topologyKey: kubernetes.io/hostname
+      weight: 100
+```
+
+> **NOTE:** Note that the Pod anti-affinity policy is `preferredDuringSchedulingIgnoredDuringExecution` as compared to `requiredDuringSchedulingIgnoredDuringExecution` ([see difference here](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity)). Therefore, in some cases a single node could have multiple NATS Pods deployed on it.
+
+
+## Test Cases
+JetStream uses a NATS optimized RAFT algorithm for clustering and high availibilty. Each stream in JetStream has a assigned stream leader NATS server. For failure scenarios, there would be two main cases:
+
+1. The k8s Node crashes where stream leader NATS Pod is deployed.
+2. The k8s Node crashes where stream leader NATS Pod is not deployed.
+
+In following sections we will run some test scenarios covering these two cases in order to check the behaviour of our system.
+
+---
+
+## Test Scenario 1: K8s Cluster Node deleted during test where stream leader NATS pod was NOT deployed
+
+### Run ID: 8/6/2022T15:26 (Duration: 5m, Event Rate: 150rps)
+
+In this test run, the k8s Node was deleted where `eventing-nats-2` Pod was deployed. The Node was deleted after 2 minutes of test execution. The `eventing-nats-2` Pod was not the stream leader.
+
+**State before test run:**
+- Stream:
+  - LastSeq# 699,293
+  - Leader: eventing-nats-1
+  - Replicas: 1
+- Consumer:
+  - Ack Floor: Stream sequence# 699,293
+
+**State after test run:**
+- Stream:
+  - LastSeq# 741,787
+  - Leader: eventing-nats-1
+  - Replicas: 1
+- Consumer:
+  - Ack Floor: Stream sequence# 741,787
+  - Redelivered Messages: 0
+  - Unprocessed Messages: 0
+
+**Test Results:**
+```
+-> Total Events Sent      by **Test Sender** : 42,494
+-> Total Events Received  by **Stream**      : 42,494 (i.e. 741,787 - 699,293) 
+-> Total Events Processed by **Consumer**    : 42,494 (i.e. 741,787 - 699,293)
+-> Total Events Received  by **Sink**        : 42,386 (There were some errors in sink for storing the data to database)
+```
+
+**Test Results Dashboard:**
+
+![](assets/ha_crash3_1.png "")
+
+**Test Sender Stats:**
+
+![](assets/ha_crash3_2.png "")
+
+
+**Finding:**
+
+The downtime of the NATS Pod (non-stream leader) did not cause any disruption in event publishing or dispatching workflow. However, it took some time for the crashed pod to come back because it takes some time for the PVC to be detached and mounted with a new K8s Node.
```

Comment:
```suggestion
The downtime of the NATS Pod (non-stream leader) did not cause any disruption in event publishing or dispatching workflow. However, it took some time for the crashed pod to come back because it takes some time for the PVC to be detached and mounted with a new Kubernetes Node.
```

Reply:
duplicate!!